crates/actors/src/epoch_service.rs

```diff
@@ -947,4 +969,182 @@ mod tests {
+
+        let mocked_block_producer = BlockProducerMockActor::mock(Box::new(move |msg, _ctx| {
+            let inner_result: eyre::Result<
+                Option<(Arc<IrysBlockHeader>, ExecutionPayloadEnvelopeV1Irys)>,
+            > = Ok(None);
+            Box::new(Some(inner_result)) as Box<dyn Any>
+        }));
+
+        let block_producer_actor_addr: Addr<BlockProducerMockActor> = mocked_block_producer.start();
+        let recipient: Recipient<SolutionFoundMessage> = block_producer_actor_addr.recipient();
+        let mocked_addr = MockedBlockProducerAddr(recipient);
+
+        let packing = Mocker::<PackingActor>::mock(Box::new(move |msg, _ctx| {
+            let packing_req = *msg.downcast::<PackingRequest>().unwrap();
+            debug!("Packing request arrived ...");
+
+            {
+                let mut lck = closure_arc.write().unwrap();
+                lck.replace(packing_req);
+            }
+
+            debug!("Packing request result pushed ...");
+            Box::new(Some(())) as Box<dyn Any>
+        }));
+
+        let vdf_steps_guard = VdfStepsReadGuard::new(Arc::new(RwLock::new(VdfState {
+            max_seeds_num: 10,
+            global_step: 0,
+            seeds: VecDeque::new(),
+        })));
+
+        let packing_addr = packing.start();
+        let mut part_actors = Vec::new();
+
+        let atomic_global_step_number = Arc::new(AtomicU64::new(0));
+
+        for sm in &storage_modules {
+            let partition_mining_actor = PartitionMiningActor::new(
+                mining_address,
+                database_provider.clone(),
+                mocked_addr.0.clone(),
+                packing_addr.clone().recipient(),
+                sm.clone(),
+                true, // do not start mining automatically
+                vdf_steps_guard.clone(),
+                atomic_global_step_number.clone(),
+            );
+
+            let part_arbiter = Arbiter::new();
+            let partition_address =
+                PartitionMiningActor::start_in_arbiter(&part_arbiter.handle(), |_| {
+                    partition_mining_actor
+                });
+            debug!("starting miner partition hash {:?}", sm.partition_hash());
+            part_actors.push(partition_address);
+        }
+
+        let submit_partition_hash = {
+            let partition_assignments_read = epoch_service_actor
+                .send(GetPartitionAssignmentsGuardMessage)
+                .await
+                .unwrap();
+            let mut maybe_partition_hash = None;
+            for (partition_hash, assignment) in
+                partition_assignments_read.read().data_partitions.iter()
+            {
+                if assignment.ledger_id == Some(Ledger::Submit.get_id()) {
+                    maybe_partition_hash = Some(partition_hash.clone());
+                    break;
+                }
+            }
+            maybe_partition_hash.expect("There should be a partition assigned to submit ledger")
+        };
```

Comment:
```suggestion
        let submit_partition_hash = {
            let partition_assignments_read = epoch_service_actor
                .send(GetPartitionAssignmentsGuardMessage)
                .await
                .unwrap();

            partition_assignments_read.read().data_partitions
                .iter()
                .find(|(_hash, assignment)| assignment.ledger_id == Some(Ledger::Submit.get_id()))
                .map(|(hash, _)| hash.clone())
                .expect("There should be a partition assigned to submit ledger")
        };
```

imo using iterators makes the code more concise and easier to follow

Reply:
Nice one ! Looks much better, almost functional programming :)